src/main/java/com/google/devtools/build/lib/remote/downloader/GrpcRemoteDownloader.java

```diff
@@ -0,0 +1,194 @@
+import java.util.TreeMap;
+
+/**
+ * A Downloader implementation that uses Bazel's Remote Execution APIs to
+ * delegate downloads of external files to a remote service.
+ *
+ * <p>See https://github.com/bazelbuild/remote-apis for more details on the
+ * exact capabilities and semantics of the Remote Execution API.
+ */
+public class GrpcRemoteDownloader implements AutoCloseable, Downloader {
+
+  private final ReferenceCountedChannel channel;
+  private final Optional<CallCredentials> credentials;
+  private final RemoteRetrier retrier;
+  private final Context requestCtx;
+  private final RemoteCacheClient cacheClient;
+  private final RemoteOptions options;
+
+  private final AtomicBoolean closed = new AtomicBoolean();
+
+  public GrpcRemoteDownloader(
+      ReferenceCountedChannel channel,
+      Optional<CallCredentials> credentials,
+      RemoteRetrier retrier,
+      Context requestCtx,
+      RemoteCacheClient cacheClient,
+      RemoteOptions options) {
+    this.channel = channel;
+    this.credentials = credentials;
+    this.retrier = retrier;
+    this.cacheClient = cacheClient;
+    this.requestCtx = requestCtx;
+    this.options = options;
+  }
+
+  @Override
+  public void close() {
+    if (closed.getAndSet(true)) {
+      return;
+    }
+    cacheClient.close();
+    channel.release();
+  }
+
+  @Override
+  public void download(
+      List<URL> urls,
+      Map<URI, Map<String, String>> authHeaders,
+      com.google.common.base.Optional<Checksum> checksum,
+      String canonicalId,
+      Path destination,
+      ExtendedEventHandler eventHandler,
+      Map<String, String> clientEnv)
+      throws IOException, InterruptedException {
+    final FetchBlobRequest request = newFetchBlobRequest(
+        options.remoteInstanceName, urls, authHeaders, checksum, canonicalId);
+    try {
+      FetchBlobResponse response = retrier.execute(
+          () -> requestCtx.call(
+              () -> fetchBlockingStub().fetchBlob(request)));
+      final Digest blobDigest = response.getBlobDigest();
+
+      retrier.execute(() -> requestCtx.call(() -> {
+        try (OutputStream out = newOutputStream(destination, checksum)) {
+          Utils.getFromFuture(cacheClient.downloadBlob(blobDigest, out));
+        }
+        return null;
+      }));
+    } catch (StatusRuntimeException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @VisibleForTesting
+  static FetchBlobRequest newFetchBlobRequest(
+      String instanceName,
+      List<URL> urls,
+      Map<URI, Map<String, String>> authHeaders,
+      com.google.common.base.Optional<Checksum> checksum,
+      String canonicalId) {
+    FetchBlobRequest.Builder requestBuilder =
+        FetchBlobRequest.newBuilder()
+            .setInstanceName(instanceName);
+    for (URL url : urls) {
+      requestBuilder.addUris(url.toString());
+    }
+    if (checksum.isPresent()) {
+      requestBuilder.addQualifiers(
+        Qualifier.newBuilder()
+            .setName("checksum.sri")
```

Comment:
Should "checksum.sri" be a constant somewhere? Do we have remote-exec specific constants anywhere?

Reply:
Changed to a constant. I can't find any common shared constants file, so I just put it at the top of this one.